File: mobile/app/src/main/java/dam/isi/frsf/utn/edu/ar/delivery/model/OrderItem.java
package dam.isi.frsf.utn.edu.ar.delivery.model;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.io.Serializable;
import java.util.List;

public class OrderItem implements Serializable {

    private final static long serialVersionUID = -1304651736295621925L;
    @SerializedName("containerType")
    @Expose
    private ContainerType containerType;
    @SerializedName("flavors")
    @Expose
    private List<Flavor> flavors = null;
    @SerializedName("sauce")
    @Expose
    private Sauce sauce;
    @SerializedName("toppings")
    @Expose
    private List<Topping> toppings = null;
    @SerializedName("quantity")
    @Expose
    private Integer quantity = 1;
    @SerializedName("delivered")
    @Expose
    private Boolean delivered;

    public ContainerType getContainerType() {
        return containerType;
    }

    public void setContainerType(ContainerType containerType) {
        this.containerType = containerType;
    }

    public OrderItem withContainerType(ContainerType containerType) {
        this.containerType = containerType;
        return this;
    }

    public List<Flavor> getFlavors() {
        return flavors;
    }

    public void setFlavors(List<Flavor> flavors) {
        this.flavors = flavors;
    }

    public OrderItem withFlavors(List<Flavor> flavors) {
        this.flavors = flavors;
        return this;
    }

    public Sauce getSauce() {
        return sauce;
    }

    public void setSauce(Sauce sauce) {
        this.sauce = sauce;
    }

    public OrderItem withSauce(Sauce sauce) {
        this.sauce = sauce;
        return this;
    }

    public List<Topping> getToppings() {
        return toppings;
    }

    public void setToppings(List<Topping> toppings) {
        this.toppings = toppings;
    }

    public OrderItem withToppings(List<Topping> toppings) {
        this.toppings = toppings;
        return this;
    }

    public Integer getQuantity() {
        return quantity;
    }

    public void setQuantity(Integer quantity) {
        this.quantity = quantity;
    }

    public OrderItem withQuantity(Integer quantity) {
        this.quantity = quantity;
        return this;
    }

    public Boolean getDelivered() {
        return delivered;
    }

    public void setDelivered(Boolean delivered) {
        this.delivered = delivered;
    }

    public OrderItem withDelivered(Boolean delivered) {
        this.delivered = delivered;
        return this;
    }

}

File: mobile/app/src/main/java/dam/isi/frsf/utn/edu/ar/delivery/model/OrderItem.kt
package dam.isi.frsf.utn.edu.ar.delivery.model

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName

data class OrderItem(
        @SerializedName("containerType")
        @Expose
        var containerType: ContainerType? = null,
        @SerializedName("flavors")
        @Expose
        var flavors: List<Flavor>? = null,
        @SerializedName("sauce")
        @Expose
        var sauce: Sauce? = null,
        @SerializedName("toppings")
        @Expose
        var toppings: List<Topping>? = null,
        @SerializedName("quantity")
        @Expose
        var quantity: Int = 1,
        @SerializedName("delivered")
        @Expose
        var delivered: Boolean = false
) : Parcelable {
    fun withContainerType(containerType: ContainerType): OrderItem {
        this.containerType = containerType
        return this
    }

    fun withFlavors(flavors: List<Flavor>): OrderItem {
        this.flavors = flavors
        return this
    }

    fun withSauce(sauce: Sauce): OrderItem {
        this.sauce = sauce
        return this
    }

    fun withToppings(toppings: List<Topping>): OrderItem {
        this.toppings = toppings
        return this
    }

    fun withQuantity(quantity: Int): OrderItem {
        this.quantity = quantity
        return this
    }

    fun withDelivered(delivered: Boolean): OrderItem {
        this.delivered = delivered
        return this
    }

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<OrderItem> = object : Parcelable.Creator<OrderItem> {
            override fun createFromParcel(source: Parcel): OrderItem = OrderItem(source)
            override fun newArray(size: Int): Array<OrderItem?> = arrayOfNulls(size)
        }
    }

    constructor(source: Parcel) : this(
            source.readParcelable<ContainerType>(ContainerType::class.java.classLoader),
            source.createTypedArrayList(Flavor.CREATOR),
            source.readSerializable() as Sauce?,
            ArrayList<Topping>().apply { source.readList(this, Topping::class.java.classLoader) },
            source.readInt(),
            1 == source.readInt()
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeParcelable(containerType, 0)
        dest.writeTypedList(flavors)
        dest.writeSerializable(sauce)
        dest.writeList(toppings)
        dest.writeInt(quantity)
        dest.writeInt((if (delivered) 1 else 0))
    }
}